Fix template forward reference

ahgittin opened the pull request:
Tests for and fixes a bug found with auto-wrapping BOMs being too strict on template resolution requirements.

aledsage commented:
@ahgittin Build failed due to no license header in `template-and-app.bom` file:
```
[ERROR] Failed to execute goal org.apache.rat:apache-rat-plugin:0.12:check (default) on project brooklyn-camp: Too many files with unapproved license: 1 See RAT report in: /home/jenkins/jenkins-slave/workspace/brooklyn-server-pull-requests/camp/camp-brooklyn/target/rat.txt -> [Help 1]
```

ahgittin commented on core/src/main/java/org/apache/brooklyn/core/typereg/BasicBrooklynTypeRegistry.java:
```diff
@@ -179,8 +180,21 @@ public RegisteredType get(String symbolicNameWithOptionalVersion) {
     public <SpecT extends AbstractBrooklynObjectSpec<?,?>> SpecT createSpec(RegisteredType type, @Nullable RegisteredTypeLoadingContext constraint, Class<SpecT> specSuperType) {
         Preconditions.checkNotNull(type, "type");
         if (type.getKind()!=RegisteredTypeKind.SPEC) { 
```
That is better, updated.

ahgittin commented:
fixed RAT and your better code style.

also added checks in other paths where unresolved might be used.  don't think any of them are necessary but doesn't hurt to try.

Graeme-Miller commented:
I have tested locally and can confirm that this fixes the issue that I saw. This PR adds a test for that issue, but thought it was best to double check.

This code looks good as well, although I don't know this area of the code particularly well.

drigodwin commented:
Thanks @ahgittin, merging now then. @Graeme-Miller was there a Jira for your issue?
